apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/LruTypePoolCache.java

```diff
@@ -0,0 +1,279 @@
+     *
+     * @param readerMode   The reader mode to use for parsing a class file.
+     */
+    public LruTypePoolCache(TypePool.Default.ReaderMode readerMode) {
+        this(readerMode, LruTypePoolCache.cacheSizeForPercentageOfCommittedHeap(
+            LruTypePoolCache.CACHE_SIZE_HALF_MB,
+            LruTypePoolCache.CACHE_SIZE_TEN_MB,
+            0.01));
+    }
+
+    /**
+     * Creates a new type locator that creates {@link TypePool}s but provides a custom {@link TypePool.CacheProvider}.
+     *
+     * @param readerMode   The reader mode to use for parsing a class file.
+     * @param maxCacheSize Determines how many {@link TypePool.Resolution}s can be cached at max.
+     */
+    public LruTypePoolCache(TypePool.Default.ReaderMode readerMode, int maxCacheSize) {
+        super(readerMode);
+        this.maxCacheSize = maxCacheSize;
+        this.sharedCache = new AtomicReference<>(new SoftReference<>(createCache()));
+        this.cacheProviders = new WeakConcurrentMap<>(false);
+    }
+
+    public static int cacheSizeForPercentageOfCommittedHeap(int minCacheSize, int maxCacheSize, double targetPercentOfHeap) {
+        long maxHeap = ManagementFactory.getMemoryMXBean().getHeapMemoryUsage().getCommitted();
+        int cacheSizeForTargetPercentOfHeap = (int) (maxHeap * targetPercentOfHeap / AVERAGE_SIZE_OF_TYPE_RESOLUTION);
+        int cacheSize = Math.max(minCacheSize, cacheSizeForTargetPercentOfHeap);
+        cacheSize = Math.min(maxCacheSize, cacheSize);
+        return cacheSize;
+    }
+
+    public LruTypePoolCache scheduleEntryEviction(final long maxAgeMs) {
+        ExecutorUtils.createSingleThreadSchedulingDaemonPool("type-cache-pool-cleaner")
+            .scheduleWithFixedDelay(new Runnable() {
+                @Override
+                public void run() {
+                    evictStaleEntries(maxAgeMs);
+                }
+            }, maxAgeMs, maxAgeMs, TimeUnit.MILLISECONDS);
+        return this;
+    }
+
+    void evictStaleEntries(long maxAgeMs) {
+        long deadline = System.currentTimeMillis() - maxAgeMs;
+        for (Iterator<Map.Entry<String, ResolutionsByClassLoader>> iterator = getSharedCache().entrySet().iterator(); iterator.hasNext(); ) {
+            Map.Entry<String, ResolutionsByClassLoader> entry = iterator.next();
+            if (entry.getValue().isExpired(deadline)) {
+                iterator.remove();
+            }
+        }
+        cacheProviders.expungeStaleEntries();
+    }
+
+    private ConcurrentMap<String, ResolutionsByClassLoader> createCache() {
+        return new ConcurrentLinkedHashMap.Builder<String, ResolutionsByClassLoader>()
+            .maximumWeightedCapacity(maxCacheSize)
+            .weigher(new EntryWeigher<String, ResolutionsByClassLoader>() {
+                @Override
+                public int weightOf(String key, ResolutionsByClassLoader value) {
+                    return Math.max(1, value.size());
```

Comment:
[question] what does weight is being used for here ? If I understand correctly that makes the entries added last to have an higher weight.

Reply:
clarified in comment

```java
// the key of the ConcurrentLinkedHashMap is the class name
// the value holds the resolutions by class loader
// if a given class name is loaded by more than one class loader,
// we need to adjust the weight of that entry accordingly
```